### ISSUE TITLE
How to do Continuous Integration

### ISSUE THREAD

**Issue by mg98** [laradock/laradock#2466]
Hello,

I need help setting up a GitHub Workflow (or should I use a service like CircleCI for that?). In the first step I would like to at least be able to run my phpunit tests. But this I need to run from inside the docker workspace, because it also needs to talk with the database. I have no clue how to start. I have little experience with CI/CD.

A small guide or advice would be very helpful.

**Comment by bestlong** [comment]
private or public repository ?
Travis-CI or CircleCI choose one first.

**Comment by mg98** [comment]
@bestlong It is a private repository. Ain't I able to do it with GitHub actions only?

**Comment by mg98** [comment]
@bestlong Hello, I would still appreciate your help a lot.

**Comment by cwilby** [comment]
@mg98 I recommend using CircleCI - the setup for that is outside the scope of Laradock.

Here's an article to get you started - in essence you need to create a `.circleci/config.yml` in the root of your project, specify the docker containers that should run (PHP/MySQL/Redis etc), then specify jobs that run in workflows.  You should add a job to build your code (install composer, run migrations), then a job to test your code (run phpunit), then a job to run a Deployer command to get the code into production.

https://medium.com/free-code-camp/laravel-continuous-deployment-with-circleci-and-deployer-c70b27fa70d2

**Comment by mg98** [comment]
Thank you very much @cwilby.

Is there a specific reason not use GitHub Actions? I understand that it is still very new and there are limited resources to it, but is there any benefit towards using CircleCI (I was always curious about that...)?

It would be so awesome to have the whole CI thing right in your GitHub repository, not having to deal with yet another service or website.

Today, after literally one month, I have actually achieved a kinda-working CI for GitHub Actions. My only problems left are persistence between jobs and that it does not cache the docker containers or images. I'm still trying to figure that out. I will share my yaml if I ever get to build a decent CI with GA...

Also you mentioned the steps `build-and-test` and `build-test-deploy`. I got confused because those look like the custom names you would give to your jobs in the CI, so I wonder where you are referring to? They are not appearing in the article.

Thanks again!!
